test/core/security/grpc_tls_crl_provider_test.cc

```diff
@@ -0,0 +1,100 @@
+//
+//
+// Copyright 2023 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//
+
+#include <grpc/support/port_platform.h>
+
+#include "src/core/lib/security/credentials/tls/grpc_tls_crl_provider.h"
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include <gtest/gtest.h>
+
+#include "absl/status/statusor.h"
+
+#include <grpc/grpc_audit_logging.h>
+#include <grpc/grpc_crl_provider.h>
+
+#include "test/core/util/test_config.h"
+#include "test/core/util/tls_utils.h"
+
+#define CRL_PATH "test/core/tsi/test_creds/crl_data/crls/ab06acdd.r0"
+#define CRL_ISSUER "/C=AU/ST=Some-State/O=Internet Widgits Pty Ltd/CN=testca"
+
+namespace grpc_core {
+namespace testing {
+
+using ::grpc_core::experimental::CertificateInfoImpl;
+using ::grpc_core::experimental::Crl;
+using ::grpc_core::experimental::CrlImpl;
+using ::grpc_core::experimental::CrlProvider;
+using ::grpc_core::experimental::StaticCrlProvider;
+
+TEST(CrlProviderTest, CanParseCrl) {
+  std::string crl_string = GetFileContents(CRL_PATH);
+  absl::StatusOr<std::shared_ptr<Crl>> result = Crl::Parse(crl_string);
+  ASSERT_TRUE(result.ok());
+  ASSERT_NE(*result, nullptr);
+  auto* crl = static_cast<CrlImpl*>(result->get());
+  EXPECT_STREQ(crl->Issuer().data(), CRL_ISSUER);
+}
+
+TEST(CrlProviderTest, InvalidFile) {
+  std::string crl_string = "INVALID CRL FILE";
+  absl::StatusOr<std::shared_ptr<Crl>> result = Crl::Parse(crl_string);
+  EXPECT_EQ(result.status(),
+            absl::InvalidArgumentError(
+                "Conversion from PEM string to X509 CRL failed."));
+}
+
+TEST(CrlProviderTest, StaticCrlProviderLookup) {
+  std::vector<std::string> crl_strings = {GetFileContents(CRL_PATH)};
+  absl::StatusOr<std::shared_ptr<CrlProvider>> result =
+      StaticCrlProvider::FromVector(crl_strings);
+  std::shared_ptr<CrlProvider> provider = std::move(*result);
```

Comment:
There's no need for this temporary variable.  See https://abseil.io/tips/181 for the preferred pattern here.

Reply:
So, I was trying to do that first, but I was getting the following error:
![image](https://github.com/grpc/grpc/assets/16962414/9c43742b-2c0e-41d1-abe6-46cb33b67783)

I'm not sure if this is some weirdness with a shared_ptr inside a `StatusOr`, or if I'm just missing something simple here?

Reply:
Calling `result->` gives you what's inside the `StatusOr<>`.  In this case, that's a `shared_ptr<>`, so you have to deference that too to get to the object.  So you can say something like `(*result)->GetCrl()`.

Note that, as per the abseil tip, you should call the variable something like `crl_provider` instead of `result`.

Please fix this pattern throughout this PR.

Reply:
Oh, duh. I believe I've fixed them throughout